File: tests/libjournal/channel_xtalk.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <j2/journal.h>
// support
#include <cassert>


// type aliases
template <bool stateV = true>
using inventory_t = pyre::journal::inventory_t<stateV>;

template <typename severityT, typename inventoryT>
using channel_t = pyre::journal::channel_t<severityT, inventoryT>;


// severity stubs
// info
class info_t :
    public channel_t<info_t, inventory_t<false>>
{
    // types
public:
    using channel_type = channel_t<info_t, inventory_t<false>>;

    // metamethods
public:
    // index initialization is required...
    info_t(const name_type &);
};

// info stub implementation
info_t::info_t(const name_type & name) :
    channel_type(name)
{}


// warning
class warning_t :
    public channel_t<warning_t, inventory_t<true>>
{
    // types
public:
    using channel_type = channel_t<warning_t, inventory_t<true>>;

    // metamethods
public:
    // index initialization is required...
    warning_t(const name_type &);
};

// warning stub implementation
warning_t::warning_t(const name_type & name) :
    channel_type(name)
{}


// error
class error_t :
    public channel_t<error_t, inventory_t<true>>
{
    // types
public:
    using channel_type = channel_t<error_t, inventory_t<true>>;

    // metamethods
public:
    // index initialization is required...
    error_t(const name_type &);
};

// error stub implementation
error_t::error_t(const name_type & name) :
    channel_type(name)
{}


// verify there is no crosstalk among the indices of different severities
int main() {
    // make a couple of info channels
    info_t info_1("info.channel_1");
    info_t info_2("info.channel_2");
    // a couple of warning channels
    warning_t warning_1("warning.channel_1");
    warning_t warning_2("warning.channel_2");
    // and a couple of error channels
    error_t error_1("error.channel_1");
    error_t error_2("error.channel_2");

    // check the states
    assert (!info_1);
    assert (!info_2);
    assert (warning_1);
    assert (warning_2);
    assert (error_1);
    assert (error_2);

    // get the info index
    const info_t::index_type & infos = info_1.index();
    // verify it has exactly two channels
    assert (infos.size() == 2);
    // one of them is "test.channel1"
    assert (infos.contains("info.channel_1"));
    // and the other is "test.channel2"
    assert (infos.contains("info.channel_2"));

    // get the warning index
    const warning_t::index_type & warnings = warning_1.index();
    // verify it has exactly two channels
    assert (warnings.size() == 2);
    // one of them is "test.channel1"
    assert (warnings.contains("warning.channel_1"));
    // and the other is "test.channel2"
    assert (warnings.contains("warning.channel_2"));

    // get the error index
    const error_t::index_type & errors = error_1.index();
    // verify it has exactly two channels
    assert (errors.size() == 2);
    // one of them is "test.channel1"
    assert (errors.contains("error.channel_1"));
    // and the other is "test.channel2"
    assert (errors.contains("error.channel_2"));

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/channel_xtalk.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# externals
import collections  # for {defaultdict}


# verify we can derive from {Channel} and create independent channel factories
def test():
    """
    Verify there is no cross talk among the indices of different channel severities
    """

    # get the package
    from j2.Channel import Channel

    # three channel subclasses
    class info(Channel):
        """
        info channel
        """
        index = collections.defaultdict(Channel.disabled_type)

    class warning(Channel):
        """
        warning channel
        """
        index = collections.defaultdict(Channel.enabled_type)

    class error(Channel):
        """
        error channel
        """
        index = collections.defaultdict(Channel.enabled_fatal_type)

    # make a couple of info channels
    info_1 = info("info.channel_1")
    info_2 = info("info.channel_2")

    # make a couple of warning channels
    warning_1 = warning("warning.channel_1")
    warning_2 = warning("warning.channel_2")

    # make a couple of error channels
    error_1 = error("error.channel_1")
    error_2 = error("error.channel_2")

    # check the states
    assert info_1.state == False
    assert info_2.state == False
    assert warning_1.state == True
    assert warning_2.state == True
    assert error_1.state == True
    assert error_2.state == True

    # get the info index
    infos = info.index
    # verify it has exactly two registered names
    assert len(infos) == 2
    # that one of them is "info.channel_1"
    assert "info.channel_1" in infos
    # and the other is "info.channel_2"
    assert "info.channel_2" in infos

    # repeat with the warning index
    warnings = warning.index
    # verify it has exactly two registered names
    assert len(warnings) == 2
    # that one of them is "warning.channel_1"
    assert "warning.channel_1" in warnings
    # and the other is "warning.channel_2"
    assert "warning.channel_2" in warnings

    # repeat with the error index
    errors = error.index
    # verify it has exactly two registered names
    assert len(errors) == 2
    # that one of them is "error.channel_1"
    assert "error.channel_1" in errors
    # and the other is "error.channel_2"
    assert "error.channel_2" in errors

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
